lib/Arch/X86/Arch.cpp

```diff
@@ -788,6 +788,54 @@ static void DecodeOperand(Instruction &inst, const xed_decoded_inst_t *xedd,
   }
 }
 
+static uint64_t BytesToCallBits(const uint8_t *bytes) {
+  // The pattern we're looking for is 5 bytes long.
+  uint64_t bits = 0;
+  bits = (bits << 8) | static_cast<uint32_t>(bytes[0]);
+  bits = (bits << 8) | static_cast<uint32_t>(bytes[1]);
+  bits = (bits << 8) | static_cast<uint32_t>(bytes[2]);
+  bits = (bits << 8) | static_cast<uint32_t>(bytes[3]);
+  bits = (bits << 8) | static_cast<uint32_t>(bytes[4]);
+  return bits;
+}
+
+static bool TryDecodeIdioms(Instruction &inst) {
+  // Check for CALL+POP idiom used to retrieve PC.
+  if (inst.bytes.size() == 6) {
+    const auto bytes = reinterpret_cast<const uint8_t *>(inst.bytes.data());
+    const auto call_bits = BytesToCallBits(bytes);
+
+    // Check that we have a CALL pointing one instruction ahead of where we are
+    // now.
+    if (call_bits != 0xe800000000) {
+      return false;
+    }
+
+    // Check whether we have a POP instruction to retrieve the return address
+    // that we just pushed to the stack.
+    Instruction pop_inst;
+    if (!inst.arch->DecodeInstruction(inst.pc + 5, &inst.bytes[5], pop_inst) ||
+        inst.function.substr(0, 3) != "POP") {
+      return false;
+    }
+
+    // We've matched the idiom. Now let's call our custom intrinsic that assigns
+    // the PC to the operand register.
+
+    // The operand should be the register that we're popping into. Just take the
+    // operand from the POP instruction.
+    assert(pop_inst.operands.size() == 1);
+    inst.operands = std::move(pop_inst.operands);
```

Comment:
- [ ] Still need to add an address calculation operand to calculate the return address of the `CALL`. So instead of assigning PC to the register, it should be PC + 5.

Reply:
[c2c5654](https://github.com/lifting-bits/remill/pull/610/commits/c2c56541b50263750348f5308ad16c246a823949)